### PR TITLE
Use FlatBuffers for the Type Registry state

### PR DESCRIPTION
**Pull request by dominiklohmann**
This changes the Type Registry state to be persisted using FlatBuffers in a versioned fashion, rather than the old CAF-serialized format. When starting VAST for the first time with this PR the old state is removed automatically.

### :memo: Checklist

- [x] All user-facing changes have changelog entries.
- [x] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

This should not be user-facing at all.

I recommend reading the code, and most importantly reviewing the new FlatBuffers table, and then giving it a spin locally. For me this has worked as expected so far.

### CODE REVIEW

File: libvast/src/system/type_registry.cpp [hunk 1]
```diff
@@ -111,27 +131,61 @@ caf::error type_registry_state::load_from_disk() {
...
+                                                "type-registry state");
+      for (const auto& [k, vs] : intermediate) {
+        auto entry = type_set{};
+        for (const auto& v : vs)
+          entry.emplace(type::from_legacy_type(v));
+        data.emplace(k, entry);
+      }
+      VAST_DEBUG("{} loaded state from disk", *self);
+      if (!std::filesystem::remove(fname, err) || err)
+        VAST_DEBUG("failed to delete legacy type-registry state");
```
Comment (dominiklohmann) [inline libvast/src/system/type_registry.cpp:158]: I wonder if before deleting we should actually persist the new state first. If that fails we shouldn't remove the old state.

Comment (dominiklohmann) [reply libvast/src/system/type_registry.cpp:161]: I considered this a sensible change and went ahead and implemented it.

File: libvast/src/system/type_registry.cpp [hunk 2]
```diff
@@ -111,27 +130,65 @@ caf::error type_registry_state::load_from_disk() {
...
+      return caf::none;
+    }
+  }
+  // Support the new FlatBuffers state.
+  {
+    const auto fname = filename();
+    const auto file_exists = std::filesystem::exists(fname, err);
+    if (err)
+      return caf::make_error(ec::filesystem_error,
+                             fmt::format("failed to find file {}: {}", fname,
```
Comment (tobim) [inline libvast/src/system/type_registry.cpp:171]: We can make this message a little clearer, the sentence can me misinterpreted to mean that the file was not found.
```suggestion
                             fmt::format("failed while trying to find file {}: {}", fname,
```

File: libvast/src/system/type_registry.cpp [hunk 3]
```diff
@@ -111,27 +130,65 @@ caf::error type_registry_state::load_from_disk() {
...
-      for (const auto& v : vs)
-        entry.emplace(type::from_legacy_type(v));
-      data.emplace(k, entry);
+  // Support the legacy CAF-serialized state, and delete it afterwards.
+  {
+    const auto fname = dir / name;
+    const auto file_exists = std::filesystem::exists(fname, err);
+    if (err)
+      return caf::make_error(ec::filesystem_error,
+                             fmt::format("failed to find file {}: {}", fname,
```
Comment (tobim) [inline libvast/src/system/type_registry.cpp:139]: We can make this message a little clearer, the sentence can me misinterpreted to mean that the file was not found.
```suggestion
                             fmt::format("failed while trying to find file {}: {}", fname,
```

**Review by tobim — APPROVED** [review]
Just a minor nit regarding an error message.

**Pull request merged**